source/extensions/path/uri_template_lib/uri_template_internal.cc

```diff
@@ -0,0 +1,386 @@
+        return absl::InvalidArgumentError("Invalid variable match");
+      }
+      var_patt = var_patt.substr(1);
+    }
+  }
+
+  return ParsedResult<Variable>(var, unconsumed_pattern);
+}
+
+absl::StatusOr<absl::flat_hash_set<absl::string_view>>
+gatherCaptureNames(struct ParsedUrlPattern pattern) {
+  absl::flat_hash_set<absl::string_view> captured_variables;
+
+  for (const ParsedSegment& segment : pattern.parsed_segments_) {
+    if (!absl::holds_alternative<Variable>(segment)) {
+      continue;
+    }
+    if (captured_variables.size() >= kPatternMatchingMaxVariablesPerUrl) {
+      return absl::InvalidArgumentError("Exceeded variable count limit");
+    }
+    absl::string_view var_name = absl::get<Variable>(segment).var_name_;
+
+    if (var_name.size() < kPatternMatchingMinVariableNameLen ||
+        var_name.size() > kPatternMatchingMaxVariableNameLen) {
+      return absl::InvalidArgumentError("Invalid variable length");
+    }
+    if (captured_variables.contains(var_name)) {
+      return absl::InvalidArgumentError("Repeated variable name");
+    }
+    captured_variables.emplace(var_name);
+  }
+
+  return captured_variables;
+}
+
+absl::Status validateNoOperatorAfterTextGlob(struct ParsedUrlPattern pattern) {
+  bool seen_text_glob = false;
+  for (const ParsedSegment& segment : pattern.parsed_segments_) {
+    if (absl::holds_alternative<Operator>(segment)) {
+      if (seen_text_glob) {
+        return absl::InvalidArgumentError("Glob after text glob.");
+      }
+      seen_text_glob = (absl::get<Operator>(segment) == Operator::TextGlob);
+    } else if (absl::holds_alternative<Variable>(segment)) {
+      const Variable& var = absl::get<Variable>(segment);
+      if (var.var_match_.empty()) {
+        if (seen_text_glob) {
+          // A variable with no explicit matcher is treated as a path glob.
+          return absl::InvalidArgumentError("Implicit variable path glob after text glob.");
+        }
+      } else {
+        for (const absl::variant<Operator, absl::string_view>& var_seg : var.var_match_) {
+          if (!absl::holds_alternative<Operator>(var_seg)) {
+            continue;
+          }
+          if (seen_text_glob) {
+            return absl::InvalidArgumentError("Glob after text glob.");
+          }
+          seen_text_glob = (absl::get<Operator>(var_seg) == Operator::TextGlob);
+        }
+      }
+    }
+  }
+  return absl::OkStatus();
+}
+
+absl::StatusOr<ParsedUrlPattern> parseURLPatternSyntax(absl::string_view url_pattern) {
+  struct ParsedUrlPattern parsed_pattern;
+
+  static const LazyRE2 printable_regex = {"^/[[:graph:]]*$"};
+  if (!RE2::FullMatch(toStringPiece(url_pattern), *printable_regex)) {
+
+    return absl::InvalidArgumentError("Invalid pattern");
+  }
+
+  // Consume the leading '/'
+  url_pattern = url_pattern.substr(1);
+
+  // Do the initial lexical parsing.
+  while (!url_pattern.empty()) {
+    ParsedSegment segment;
+    if (url_pattern[0] == '*') {
+
```

Comment:
nit: remove all these empty lines after the ifs.

Reply:
Done